src/CodeGen_X86.cpp

```diff
@@ -306,10 +306,249 @@ void CodeGen_X86::visit(const Select *op) {
 
 }
 
+bool CodeGen_X86::try_visit_float16_cast(const Cast* op) {
+    Type destTy = op->type;
+    Type srcTy = op->value.type();
+
+    // half -> single/double
+    if (destTy.is_float() && srcTy.is_float() && srcTy.bits == 16 && destTy.bits > 16) {
+        internal_assert(destTy.bits == 64 || destTy.bits == 32) << "Unexpected float type\n";
+
+        if (target_needs_software_cast_from_float16_to(destTy)) {
+            // Let parent class codegen calling software implementation of cast
+            return false;
+        }
+
+        // Use @llvm.x86.vcvtph2ps.128 or @llvm.x86.vcvtph2ps.256 to handle the
+        // conversion
+        internal_assert(srcTy.width == destTy.width) << "Source and destination widths must match\n";
+        if (srcTy.width <= 4) {
+            // Use @llvm.x86.vcvtph2ps.128. This returns <4 x float>
+            // FIXME: Codegen using other widths (e.g. 3) seems to give wrong
+            // results explicitly disable until this can be investigated.
+            internal_assert(srcTy.width == 4 || srcTy.width == 1) <<
+              "FIXME: Doing codegen with width less than four produces incorrect results\n";
+
+            // Codegen argument
+            llvm::Value* valueToCast = codegen(op->value);
+
+            // Note with width is 8 here because that is what is expected by
+            // the instrinsic. Only the first 4 lanes matter here.
+            // Slice to correct width
+            if (srcTy.width == 1) {
+                // Handle scalar value by inserting as the first element
+                // in a vector
+                internal_assert(!(valueToCast->getType()->isVectorTy()));
+                Constant* zeroHalfVectorSplat = ConstantFP::get(f16x8, 0.0);
+                valueToCast = builder->CreateInsertElement(/*Vec=*/zeroHalfVectorSplat,
+                                                           /*NewElt=*/valueToCast,
+                                                           /*Idx=*/ConstantInt::get(i32, 0)
+                                                          );
+            } else {
+                // Handle vectorized inputs by extending to correct width.
+                // Introduced undef lanes will be removed later.
+                valueToCast = slice_vector(valueToCast, /*start=*/0, /*size=*/8);
+            }
+
+            // Cast argument to <8 x i16> as expected by the intrinsic
+            valueToCast = builder->CreateBitCast(valueToCast, i16x8);
+
+            // Can't use call_intrin() because it asserts the number of elements
```

Comment:
I think there's a much simpler implementation to be had here. Codegen for vectorized call nodes already pads things out to native vectors, or slices things up into native vectors, or scalarizes. It does this according to what functions are found to exist in the runtime. E.g. if only a scalar function exists, it uses that on each lane. If a vector function exists, it pads out the vectors as needed and calls that. See the CodeGen_LLVM.cpp around line 2708

So what you could do is just always codegen a Call node of halide_float16_bits_to_float, and then in runtime/x86_f16c.ll you provide a function halide_float16_bits_to_float which has the right ll for the hardware conversion, and functions with suffixes like halide_float16_bits_to_floatx8, etc which do the vectorized versions. You'd need to rejigger get_initial_module_for_target to include x86_f16c.ll instead of float16_t.cpp for x86 with f16c.

This way new platforms that support float16 in hardware only need to provide implementations of halide_float16_bits_to_float and the like in their runtime modules, and set up get_initial_module_for_target to use those runtime modules, and any vector versions will get used as appropriate automatically.

For an example of how this works, see that floor_f32x4 exists in x86_sse41.ll, and floor_f32x8 exists in x86_avx.ll. codegen of a Call node to floor_f32 of some arbitrary vector width uses whichever ones exist intelligently.